File: DSA/graph/graph_print.py
def depth_first_print(graph, source):
    print(source)
    for neighbor in graph[source]:
        depth_first_print(graph, neighbor)

def breadth_first_print(graph, source):
    queue = [source]

    while len(queue) > 0:
        current = queue.pop(0)
        print(current)
        for neighbor in graph[current]:
            queue.append(neighbor)


if __name__ == '__main__':
    graph = {
        'a': ['b', 'c'],
        'b': ['d'],
        'c': ['e'],
        'd': ['f'],
        'e': [],
        'f': [],
    }
    print('Depth first print')
    depth_first_print(graph, 'a')
    print('Breadth first print')
    breadth_first_print(graph, 'a')

File: DSA/graph/graphPrint.js
// const depthFirstPrint = (graph, source) => {
// 	const stack = [source]
// 	while (stack.length > 0) {
// 		let current = stack.pop()
// 		console.log(current)
// 		for (let neighbor of graph[current]) {
// 			stack.push(neighbor)
// 		}
// 	}
// }

// recursion 
const depthFirstPrint = (graph, source) => {
	console.log(source)
	for (let neighbor of graph[source]) {
		depthFirstPrint(graph, neighbor)
	}
}

const breadthFirstPrint = (graph, source) => {
	const queue = [source]
	while (queue.length > 0) {
		let current = queue.shift()
		console.log(current)
		for (let neighbor of graph[current]) {
			queue.push(neighbor)
		}
	}
}

const graph = {
	a: ['b', 'c'],
	b: ['d'],
	c: ['e'],
	d: ['f'],
	e: [],
	f: [],
}

console.log('Depth first traversal')
depthFirstPrint(graph, 'a')

console.log('Breadth first traversal')
breadthFirstPrint(graph, 'a')

